### ISSUE TITLE
Refactor: Add label to embedder registration entry and embedder interface

### ISSUE THREAD

**Issue by markvilar** [markvilar/viper#79]
## 1. Problem and motivation

The `key` used to register and look up embedders (`EmbedderRegistrationEntry.key` in `src/viper/registry.py`) is formatted for lookup and mapping, not for presentation — it is unsuitable for use directly in plots, visualizations, and other human-facing output. There is currently no dedicated presentation string carried alongside the registration entry or the `ImageEmbedder` protocol.

The `ImageEmbedder` protocol (`src/viper/types.py`) does expose a `name` property, but "name" is ambiguous with respect to the `key`/label distinction and should be renamed to `label` for consistency with the registration entry and with the existing `label` precedent in `src/viper/forge/registry.py`. We would also like to consider adding a `key` property to `ImageEmbedder`, so that an embedder instance carries both its registry `key` and its presentation `label`, mirroring the registration entry and letting downstream code (e.g. output naming, plotting, logging) reliably access either the lookup identifier or the display string from the embedder itself.

### 2.1 Registration entry

Add a `label: str` field to `EmbedderRegistrationEntry`, alongside the existing `key` and `family` fields:

```python
@dataclass(frozen=True)
class EmbedderRegistrationEntry:
    """A declarative registration for a remote-URL-backed embedder."""

    key: str
    """Flat lookup key for the variant (e.g. "megaloc-512d-svd-truncated")."""

    label: str
    """Presentation string for the variant (e.g. "MegaLoc (512d, SVD-truncated)")."""

    family: str
    """Grouping label for related keys (e.g. "megaloc"); enumeration only."""

    checkpoint_url: str
    """URL the variant is built from (state-dict checkpoint, hub repo, ...)."""

    factory: Callable[[str], ImageEmbedder]
    """Builds the embedder from the URL; owns the load strategy end to end."""
```

`register_embedder_factory` gains a matching `label` keyword and a parallel `_embedder_labels: dict[str, str]` map, so `key`-only registrations (the `@register_embedder_factory(...)` decorator used directly by model modules) also carry a presentation string, not just entries built through `register_embedder_entries`.

`family` remains a grouping concept for enumeration; `label` is a distinct, per-key presentation string. A variant's `label` will often incorporate its `family` (e.g. all MegaLoc variants share the "MegaLoc" root) but the two must stay independently settable, since not every distinguishing detail belongs in the grouping key.

### 2.2 Image embedder interface

Rename `ImageEmbedder.name` to `label`, and add a `key` property:

```python
class ImageEmbedder(typing.Protocol):
    """Class representing the interface for an image embedder."""

    @property
    def key(self) -> str:
        """Returns the registry lookup key of the embedder."""
        ...

    @property
    def label(self) -> str:
        """Returns the presentation label of the embedder."""
        ...

    @property
    def vector_size(self) -> int:
        ...

    @property
    def embedder_parameters(self) -> dict[str, typing.Any]:
        ...

    @property
    def device(self) -> str:
        ...

    def __call__(self, images: torch.Tensor) -> torch.Tensor: ...
    def forward(self, images: torch.Tensor) -> torch.Tensor: ...
```

With `key` and `label` both on the interface, any `ImageEmbedder` instance can identify itself for both lookup and presentation without a caller needing to consult the registry separately — useful when an embedder has been passed around detached from its registration entry (e.g. into a plotting routine).

### 2.3 Model-specific implementations

Every model wrapper (`AnyLocWrapper`, and the NetVLAD/CliqueMining/SALAD/CosPlace/MixVPR/EigenPlaces/MegaLoc equivalents) currently hardcodes its own `name` property. We add `key`/`label` as ordinary constructor parameters and properties, matching the rest of the wrapper's fields:

```python
class AnyLocWrapper(torch.nn.Module):
    """
    Class representing a wrapper for the AnyLoc model.
    """

    def __init__(self, impl: torch.nn.Module, key: str, label: str) -> None:
        """Initializer method."""
        super().__init__()
        self._impl = impl
        self._key = key
        self._label = label
        # NOTE: Add dummy parameter to infer device
        self._param = torch.nn.Parameter(torch.tensor(1.0))

    @property
    def key(self) -> str:
        """Returns the registry lookup key of the embedder."""
        return self._key

    @property
    def label(self) -> str:
        """Returns the presentation label of the embedder."""
        return self._label

    @property
    def vector_size(self) -> int:
        ...
```

This is more boilerplate per wrapper than a shared wrapper class would need, but it keeps each model's constructor self-contained and legible as a reference for what a model implementation requires — `key`/`label` sit alongside the model's other required fields rather than being bolted on after construction. The values themselves are never hardcoded inside the wrapper; they're always supplied by the caller (the factory), so a given wrapper class stays reusable across every `key`/`label` pair a factory wants to construct it with.

### 2.4 Model-specific factories

Factories become the layer responsible for supplying `key`/`label` to the model constructor. `EmbedderRegistrationEntry.factory` changes shape to accept them alongside the checkpoint URL:

```python
@dataclass(frozen=True)
class EmbedderRegistrationEntry:
    key: str
    label: str
    family: str
    checkpoint_url: str
    factory: Callable[[str, str, str], ImageEmbedder]
    """Builds the embedder from (url, key, label); owns the load strategy end to end."""
```

```python
def _load_megaloc_from_url(url: str, key: str, label: str) -> ImageEmbedder:
    """Loads a MegaLoc variant from a state-dict checkpoint URL. Requires CUDA."""
    if not torch.cuda.is_available():
        raise RuntimeError("CUDA is required for MegaLoc but is not available.")
    state_dict = torch.hub.load_state_dict_from_url(
        url, map_location="cpu", weights_only=True
    )
    return build_megaloc_from_state_dict(state_dict, key=key, label=label).eval().cuda()
```

```python
_ENTRIES = [
    EmbedderRegistrationEntry(
        key="megaloc",
        label="MegaLoc",
        family=_FAMILY,
        checkpoint_url=f"{_RELEASES}/megaloc-8448d-v1/megaloc-8448d-v1.pth",
        factory=_load_megaloc_from_url,
    ),
    EmbedderRegistrationEntry(
        key="megaloc-512d-svd-truncated",
        label="MegaLoc 512D SVD-Truncated",
        family=_FAMILY,
        checkpoint_url=(
            f"{_RELEASES}/megaloc-svd-truncated-v1.0/megaloc-512d-svd-truncated-v1.pth"
        ),
        factory=_load_megaloc_from_url,
    ),
    # ...
]
```

`register_embedder_entries` calls `entry.factory(entry.checkpoint_url, entry.key, entry.label)`, so the same shared factory function correctly produces a differently labeled embedder per entry. The single-entry, decorator-based factories (AnyLoc and similar) follow the same shape — `register_embedder_factory` gains a `label` keyword and passes `key`/`label` through to the decorated function:

```python
@register_embedder_factory(key="anyloc", label="AnyLoc", family="anyloc")
def load_anyloc(key: str, label: str) -> ImageEmbedder:
    """Loads an AnyLoc model."""
    if not torch.cuda.is_available():
        raise RuntimeError("CUDA is required for AnyLoc but is not available.")

    impl: torch.nn.Module = torch.hub.load(
        "AnyLoc/DINO", "get_vlad_model", backbone="DINOv2",
        domain="unstructured", device="cuda",
    )
    return AnyLocWrapper(impl=impl, key=key, label=label).eval()
```

This puts the `key`/`label` plumbing entirely in the factory layer — the piece of code whose job is already to know how to construct a specific, fully-configured embedder — while every model wrapper class stays a plain, reusable constructor with no implicit state and no post-construction setup step required before it's safe to use.

### 2.5 Registration entries to update

Of the current embedder registrations, only MegaLoc is a checkpoint-hosted-as-GitHub-release entry built through `EmbedderRegistrationEntry`/`register_embedder_entries` — the rest (AnyLoc, NetVLAD, CliqueMining, SALAD, CosPlace, MixVPR, EigenPlaces) still use the older `torch.hub`-based, decorator-only registration and haven't yet been migrated to the checkpoint-as-GitHub-asset approach. This issue only updates the four MegaLoc entries; the other models are out of scope here and will pick up `key`/`label` when they're separately migrated to the new registration approach.

```python
_ENTRIES = [
    EmbedderRegistrationEntry(
        key="megaloc",
        label="MegaLoc",
        family=_FAMILY,
        checkpoint_url=f"{_RELEASES}/megaloc-8448d-v1/megaloc-8448d-v1.pth",
        factory=_load_megaloc_from_url,
    ),
    EmbedderRegistrationEntry(
        key="megaloc-256d-svd-truncated",
        label="MegaLoc 256D SVD-Truncated",
        family=_FAMILY,
        checkpoint_url=(
            f"{_RELEASES}/megaloc-svd-truncated-v1.0/megaloc-256d-svd-truncated-v1.pth"
        ),
        factory=_load_megaloc_from_url,
    ),
    EmbedderRegistrationEntry(
        key="megaloc-512d-svd-truncated",
        label="MegaLoc 512D SVD-Truncated",
        family=_FAMILY,
        checkpoint_url=(
            f"{_RELEASES}/megaloc-svd-truncated-v1.0/megaloc-512d-svd-truncated-v1.pth"
        ),
        factory=_load_megaloc_from_url,
    ),
    EmbedderRegistrationEntry(
        key="megaloc-1024d-svd-truncated",
        label="MegaLoc 1024D SVD-Truncated",
        family=_FAMILY,
        checkpoint_url=(
            f"{_RELEASES}/megaloc-svd-truncated-v1.0/megaloc-1024d-svd-truncated-v1.pth"
        ),
        factory=_load_megaloc_from_url,
    ),
]
```

## 3. Design decisions

| Decision | Reasoning |
|---|---|
| Rename `ImageEmbedder.name` to `label`, and add a `key` property | `name` was ambiguous with respect to the `key`/presentation-string distinction; `label` matches the existing precedent in `src/viper/forge/registry.py` and the new `EmbedderRegistrationEntry.label` field. Adding `key` lets an embedder instance identify itself for lookup even when detached from its registration entry. |
| Add `label: str` as its own field on `EmbedderRegistrationEntry`, distinct from `family` | `family` is a grouping concept used only for enumeration; `label` is a presentation string. Collapsing them would force a grouping key to also serve as display text, which doesn't hold for variants that share a family but need distinct labels (e.g. the MegaLoc SVD-truncated variants). |
| Add `key`/`label` as ordinary constructor parameters on each model wrapper (Section 2.3, "option a"), rather than a shared wrapper/decorator class | Keeps every model wrapper class self-contained and legible as a reference implementation, with no implicit post-construction setup step and no extra object layer between the model and the registry. Trades a small amount of boilerplate per wrapper for that clarity. |
| Thread `key`/`label` through the factory function signature (Section 2.4, "option 1"), rather than injecting them into the constructed embedder afterwards | Keeps embedders fully immutable and valid immediately upon construction — no window where an embedder exists without its `key`/`label` set. Puts the plumbing in the factory layer, whose job is already to know how to construct one specific, fully-configured embedder, rather than in the model implementation. |
| Scope this issue to the four MegaLoc registration entries only (Section 2.5) | The other models (AnyLoc, NetVLAD, CliqueMining, SALAD, CosPlace, MixVPR, EigenPlaces) still use the older `torch.hub`-based, decorator-only registration and haven't yet been migrated to the checkpoint-as-GitHub-asset approach that `EmbedderRegistrationEntry` targets. They'll pick up `key`/`label` when separately migrated, rather than mixing that migration into this change. |

## 4. Open questions

*None outstanding.* All previously open questions have been settled:

- **Label formatting conventions.** Resolved: labels follow `"<Model> <Config>"`, e.g. `"MegaLoc"`, `"MegaLoc 512D SVD-Truncated"` (see Section 2.5).
- **Uniqueness of `label`.** Resolved: not enforced. `key` remains the unique identifier; duplicate labels are permitted.
- **Downstream call sites of `ImageEmbedder.name`.** Resolved: a repo-wide scan for `.name` usage on embedder instances will be done as part of implementing this issue, before the rename lands.
- **Fate of `family`.** Resolved: left as-is for now; not part of this issue's scope.

## 5. Notes

- The other model implementations (AnyLoc, NetVLAD, CliqueMining, SALAD, CosPlace, MixVPR, EigenPlaces) are out of scope here (see Section 2.5) — they'll need a follow-up issue to migrate them to the checkpoint-as-GitHub-asset registration approach and adopt `key`/`label`.
- Implementation should start with a repo-wide search for `.name` usage on embedder instances, to scope the blast radius of the `name` → `label` rename before making the change.
- Depends on nothing outstanding; can be implemented independently of other open issues.
- The `ImageEmbedder.name` → `label` rename (plus the new `key` property) is a breaking public API change. The latest released version is v0.5.0; once this issue lands, the version should be bumped to at least v0.6.0.

- The README's code blocks reference the pre-refactor interface (`embedder.name`, the `ImageEmbedder` protocol summary listing `name: str`, and the custom-embedder example implementing `name`/`device -> str`). These need to be updated to the new `key`/`label` properties and `device -> torch.device` return type as part of this issue, so the README doesn't drift from the actual interface.
